Lib/email/generator.py

```diff
@@ -186,8 +186,10 @@ def _write(self, msg):
         # If we munged the cte, copy the message again and re-fix the CTE.
         if munge_cte:
             msg = deepcopy(msg)
-            msg.replace_header('content-transfer-encoding', munge_cte[0])
-            msg.replace_header('content-type', munge_cte[1])
+            if msg.get('content-transfer-encoding') is not None:
+                msg.replace_header('content-transfer-encoding', munge_cte[0])
+            if msg.get('content-type') is not None:
+                msg.replace_header('content-type', munge_cte[1])
```

Comment:
The CTE is munged only when (a) we are translating to ASCII-only strings (Generator and not BytesGenerator) and (b) we have a non-None 'charset', which means we *have* a content-type header.  So the content-type header should be set unconditionally so that if someone changes (b) we might notice.  Which means we should also have a test that has non-ascii and neither header...which might find a new error :)

Reply:
The original bug report https://bugs.python.org/issue27321 was for a message which had a Content-Type: but no Content-Transfer-Encoding:.
Based on @bitdancer  comment, perhaps the code should be:
```
            msg.replace_header('content-type', munge_cte[1])
            if msg.get('content-transfer-encoding') is not None:
                msg.replace_header('content-transfer-encoding', munge_cte[0]
            else:
                msg['Content-Transfer-Encoding'] = munge_cte[0]
```
Of course, a test with a non-ascii message with neither header wouldn't hurt in any case.

Reply:
Yes that looks good.